---
title: "nf-core/sarek paper"
subtitle: "Harder, Better, Faster, Stronger"
pubDate: 2024-04-25T00:19:00+01:00
headerImage: "/assets/images/blog/sarek/skierfe.jpg"
headerImageAlt: "Mount Skierfe outlined showing the inspiration for the Sarek logo"
authors:
    - "FriederikeHanssen"
    - "maxulysse"
label:
    - "pipeline"
embedHeaderImage: true
---

import { YouTube } from "@astro-community/astro-embed-youtube";

import work_pic1 from "@assets/images/blog/core_retreat_Feb_2024/work_pic1.png";
import sarek_subway from "@assets/images/blog/sarek/sarek_subway.png";
import cram_meme from "@assets/images/blog/sarek/cram_meme.jpg";
import cram_graph from "@assets/images/blog/sarek/storage_preprocessing_reduced_cumulative.png";
import dataflow from "@assets/images/blog/sarek/paper_dataflow.png";
import illumina from "@assets/images/blog/sarek/benchmarking_snvs_combination.png";
import not_illumina from "@assets/images/blog/sarek/benchmarking_snps_combination_nonillumina.png";
import contributors from "@assets/images/blog/sarek/contributors.png";

import { Image } from "astro:assets";

We are extremely happy to see this [paper](https://academic.oup.com/nargab/article/6/2/lqae031/7658070) out describing the changes and
updates to [nf-core/sarek](https://github.com/nf-core/sarek/), the DNA variant calling pipeline, in the last several years.

In 2020, we embarked on the journey of rewriting the whole pipeline in DSL2. One of the major motivations was to bring down
cloud computing costs and generally reduce storage space and computational resources.

# Overview

No nf-core pipeline without a metro map ðŸš‡:

<Image src={sarek_subway} alt="Metromap of nf-core/sarek" height="700" densities={[1.5, 2]} />

# New tools

We added new tools: BwaMem2 and DragMap for alignment, more variant callers (DeepVariant, GATK HaplotypeCaller Joint Calling &
Single Sample variant recalibration, CNVKit, Tiddit), and more annotation possibilities.

Some tools were replaced: Trimming is now done with FastP, CRAM quality control with Mosdepth. For convenience we added more
quality control options: When starting from variant-calling directly, all input files can now run through the alignment QC steps.

![Thor throwing a cup and screaming 'Another'](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExMDMycW55dDFmNHNzODllanVxbndzZzA4eHZlNmM1NWQ0anFhZnFkMyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/ziLadIVnOGCKk/giphy.gif)

# Resource optimization

## Use CRAM files

We ditched BAM format where possible and switched to CRAM saving us 4x work storage space.

<Image src={cram_meme} alt="Toy story meme" height="300" densities={[1.5, 2]} />
<Image src={cram_graph} alt="Bar chart showing reduction in storage space usage" height="400" densities={[1.5, 2]} />

## Split files (but not too much)

The `splitFastq()` operator was replaced by a FastP process to split the Fastq files before
alignment (default 12) plus replacing trimgalore! for read trimming.
We also changed the default grouping of the intervals for BQSR to 21 (instead of 124) reducing storage space another 4x and speeding up processing.

    <Image src={dataflow} alt="Evaluation of FastP usage and different interval group sizes for BQSR" height="500" densities={[1.5, 2]} />

# Cost savings

Overall, we reduced computational costs on AWS (last summer, using spot instances)
by 70% to about $20 from FASTQs to annotated VCFs using Strelka, Manta, and VEP.

![Little geco saying: 'you could save'](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExcnUwZnV4eGkycjR5b3R3M21wbjNlN2tkZGdtbXJqam1tbmYxNGZxZCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/AlRadeCzctWmA8fnNG/giphy.gif)

# Benchmarking: a.k.a is it any good?

We benchmarked the germline track with Illumina, MGI, and BGI GiaB samples and the somatic track with Seq2C samples.
We recently joined the NCBench effort to continuously validate the pipeline on release.

<Image src={illumina} alt="" height="385" densities={[1.5, 2]} />
<Image src={not_illumina} alt="" height="385" densities={[1.5, 2]} />

# Team work makes dream work

This was a gigantic team effort with [Lasse Folkersen](https://github.com/lassefolkersen), [Anders Sune Pedersen](https://github.com/asp8200), [Francesco Lescai](https://github.com/lescai), [
Susanne Jodoin](https://github.com/SusiJo), [Edmund Miller](https://github.com/Edmundmiller), Matthias Seybold, [Oskar Wacker](https://github.com/WackerO),
[Nick Smith](https://github.com/nickhsmith), [Gisela Gabernet](https://github.com/ggabernet), [Sven Nahnsen](https://uni-tuebingen.de/forschung/forschungsinfrastruktur/zentrum-fuer-quantitative-biologie-qbic/team/prof-dr-sven-nahnsen/),
and many many others from the nf-core community:

    <Image src={contributors} alt="" height="500" densities={[1.5, 2]} />

Also shout out to all the amazing people starting sarek way back in 2016:
Szilveszter Juhos, Malin Larsson, Pall I. Olason, Marcel Martin, Jesper Eisfeldt,
Sebastian DiLorenzo, Johanna Sandgren, Teresita DÃ­az De StÃ¥hl, Phil Ewels,
Valtteri Wirta, Monica NistÃ©r, Max KÃ¤ller, and BjÃ¶rn Nystedt.

# Join the fun

If you want to join us, visit: https://nf-co.re/join/ weâ€™re on the #sarek channel on slack,
and youâ€™re welcome to join #sarek_dev if you really want to get involved.

# There is more

If you want to know more, here are some recent talks detailing the changes and development journey:

    <YouTube id="https://www.youtube.com/watch?v=pfxEJJs63y8" poster='https://i.ytimg.com/vi/pfxEJJs63y8/maxresdefault.jpg'/> <YouTube id="https://www.youtube.com/watch?v=ivHBfilnUIY" poster='https://i.ytimg.com/vi/ivHBfilnUIY/maxresdefault.jpg' />
