---
title: "Bytesize: nf-core/variantbenchmarking"
subtitle: Kübra Narci, German Human Genome-Penome Archieve, DKFZ Heidelberg
type: talk
startDate: "2025-04-01"
startTime: "13:00+01:00"
endDate: "2025-04-01"
endTime: "13:30+01:00"
youtubeEmbed: https://youtu.be/nhrm3M1Fsg8
embedAt: "variantbenchmarking"
locations:
  - name: Online
    links:
      - https://youtu.be/nhrm3M1Fsg8
---

Definitively not an Aprils fool joke! The nf-core pipeline [nf-core/varianbenchmarking](https://nf-co.re/variantbenchmarking/1.0.0/) had it's first release and this week Kübra ([@kubranarci](https://github.com/kubranarci)) is going to talk about its use cases and features!

nf-core/variantbenchmarking is designed to evaluate and validate the accuracy of variant calling methods in genomic research.
The pipeline is primarily tuned well for available gold standard truth sets like Genome in a Bottle or SEQC2 but it can be used to compare any two variant calling results.
The workflow provides benchmarking tools for small variants including SNVs and INDELs, Structural Variants (SVs) and Copy Number Variations (CNVs) for germline and somatic analysis.
